ext/app/electron/config.ts

```diff
@@ -0,0 +1,241 @@
+import * as electron from "electron";
+import * as fse from "fs-extra";
+import * as ini from "ini";
+import * as log from "app/server/lib/log";
+import * as net from 'net';
+import * as os from "os";
+import * as packageJson from "desktop.package.json";
+import * as path from "path";
+import bluebird from 'bluebird';
+import { commonUrls } from "app/common/gristUrls";
+
+const APPDATA_DIR = (["win32", "darwin"].includes(process.platform))
+  ? path.join(electron.app.getPath("appData"), packageJson.name)
+  : path.join(electron.app.getPath("home"), ".local", "share", packageJson.name);
+
+// Electron's app.getPath("userData") uses productName instead of name
+// but productName should be "full capitalized name", not ideal for naming our config directory.
+const DEFAULT_CONFIG_FILE = path.join(APPDATA_DIR, "config.ini");
+const NO_VALIDATION = () => true;
+
+/**
+ * Suggest a value for an environment variable. If the variable is already set, do nothing. Otherwise set the value.
+ * @param name The name of the environment variable.
+ * @param value The value to suggest.
+ */
+function suggestEnv(name: string, value: string): void {
+  if (process.env[name] === undefined) {
+    process.env[name] = value;
+  }
+}
+
+/**
+ * Copied from grist-core, since it is unsafe to import core code at this point.
+ */
+async function getAvailablePort(firstPort: number = 8000, optCount: number = 200): Promise<number> {
+  const lastPort = firstPort + optCount - 1;
+  async function checkNext(port: number): Promise<number> {
+    if (port > lastPort) {
+      throw new Error("No available ports between " + firstPort + " and " + lastPort);
+    }
+    return new bluebird((resolve: (p: number) => void, reject: (e: Error) => void) => {
+      const server = net.createServer();
+      server.on('error', reject);
+      server.on('close', () => resolve(port));
+      server.listen(port, 'localhost', () => server.close());
+    })
+    .catch(() => checkNext(port + 1));
+  }
+  return bluebird.try(() => checkNext(firstPort));
+}
+
+// The ini library recognizes boolean values, but not numbers. All other values are treated as strings.
+type INI = { [key: string]: (INI | string | boolean) }
+
+class Config {
+
+  #config: INI;
```

Comment:
Given that this is typescript, something like `private _config: INI` would be more compatible with the style in the rest of Grist.

Or better, do the declaration in the constructor, like:

```
public constructor(private _config: INI) {
}
```

That would be even more like the style elsewhere.

Reply:
I did a quick search and learned that it was "safer" to use `#` since it ensures the field remains inaccessible in the compiled js, so I went with `#`. Should I disregard this concern and always use the ts `private` syntax?

Reply:
Yes please, can you use ts `private`? Compile time privacy is all we need, there's no threat model here, just marking for developers and tsc that this field is an implementation detail.